onnxruntime/core/session/model_editor_c_api.cc

```diff
@@ -0,0 +1,358 @@
+#include "core/session/abi_session_options_impl.h"
+#include "core/session/inference_session.h"
+#include "core/session/model_editor_api.h"
+#include "core/session/ort_apis.h"
+#include "core/session/ort_env.h"
+#include "core/session/utils.h"
+
+using namespace onnxruntime;
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::CreateValueInfo, _In_ const char* name, _In_ const OrtTypeInfo* type_info,
+                    _Outptr_ OrtValueInfo** value_info) {
+  API_IMPL_BEGIN
+  if (name == nullptr || *name == '\0') {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "name cannot be null or empty string");
+  }
+
+  if (type_info == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "type_info cannot be null");
+  }
+
+  if (type_info->type != ONNX_TYPE_TENSOR) {
+    return OrtApis::CreateStatus(ORT_FAIL, "Only tensor types are supported currently");
+  }
+
+  if (type_info->tensor_type_info == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "tensor_type_info cannot be null");
+  }
+
+  auto vi = std::make_unique<OrtValueInfo>();
+  vi->name = name;
+  vi->type_info = type_info->Clone();
+
+  *value_info = vi.release();
+
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::CreateNode, const char* operator_name, const char* domain_name,
+                    _In_ const char* node_name,
+                    _In_reads_(input_names_len) const char* const* input_names, size_t input_names_len,
+                    _In_reads_(output_names_len) const char* const* output_names, size_t output_names_len,
+                    _In_reads_(attribs_len) _Inout_opt_ OrtOpAttr** attributes, _In_opt_ size_t attribs_len,
+                    _Outptr_ OrtNode** node) {
+  API_IMPL_BEGIN
+  auto n = std::make_unique<OrtNode>();
+  n->operator_name = operator_name;
+  n->domain_name = domain_name == kOnnxDomainAlias ? kOnnxDomain : domain_name;
+  n->node_name = node_name;
+
+  n->input_names.reserve(input_names_len);
+  for (size_t i = 0; i < input_names_len; ++i) {
+    n->input_names.push_back(input_names[i]);
+  }
+
+  n->output_names.reserve(output_names_len);
+  for (size_t i = 0; i < output_names_len; ++i) {
+    n->output_names.push_back(output_names[i]);
+  }
+
+  if (attributes != nullptr) {
+    n->attributes.reserve(attribs_len);
+    for (size_t i = 0; i < attribs_len; ++i) {
+      n->attributes.push_back(*reinterpret_cast<const ONNX_NAMESPACE::AttributeProto*>(attributes[i]));
+      // take ownership. as we took a copy that means releasing the original value
+      OrtApis::ReleaseOpAttr(attributes[i]);
+      attributes[i] = nullptr;
+    }
+  }
+
+  *node = n.release();
+  return nullptr;
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtModelEditorAPI::CreateGraph, _Outptr_ OrtGraph** graph) {
+  API_IMPL_BEGIN
+  auto g = std::make_unique<OrtGraph>();
+
+  // do some reserves to reduce reallocation. if we had a hint about sizes upfront that would be optimal
+  g->inputs.reserve(8);
```

Comment:
>g->inputs.reserve(8); [](http://example.com/codeflow?start=2&length=21)

I think it would not be necessary if we simply used InlinedVector since the type internals are not exposed anyway.